modules/core/src/test/java/org/apache/ignite/internal/IgniteClientReconnectCustomBinaryContextTest.java

```diff
@@ -0,0 +1,296 @@
+                objectField.setAccessible(true);
+
+                typeField = ReflectionUtils.findField(cls, "type");
+                typeField.setAccessible(true);
+            }
+            catch (ClassNotFoundException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        @Override public void writeBinary(Object obj, BinaryWriter writer) throws BinaryObjectException {
+            Byte type = (Byte)ReflectionUtils.getField(typeField, obj);
+            writer.writeByte("type", type);
+            if (type == DURATION_TYPE) {
+                Duration duration = (Duration)ReflectionUtils.getField(objectField, obj);
+                writer.writeString("duration", duration != null ? duration.toString() : null);
+            }
+            else {
+                Temporal date = (Temporal)ReflectionUtils.getField(objectField, obj);
+                writer.writeInt("year", date.get(YEAR));
+                writer.writeInt("month", date.get(MONTH_OF_YEAR));
+                writer.writeInt("day", date.get(DAY_OF_MONTH));
+                if (type == LOCAL_DATE_TIME_TYPE) {
+                    writer.writeInt("hour", date.get(HOUR_OF_DAY));
+                    writer.writeInt("minute", date.get(MINUTE_OF_HOUR));
+                    writer.writeInt("second", date.get(SECOND_OF_MINUTE));
+                    writer.writeInt("nano", date.get(NANO_OF_SECOND));
+                }
+            }
+        }
+
+        @Override public void readBinary(Object obj, BinaryReader reader) throws BinaryObjectException {
+            byte type = reader.readByte("type");
+
+            if (type == DURATION_TYPE) {
+                String string = reader.readString("duration");
+                Duration duration = string != null ? Duration.parse(string) : null;
+                ReflectionUtils.setField(objectField, obj, duration);
+
+            }
+            else {
+                int year = reader.readInt("year");
+                int month = reader.readInt("month");
+                int day = reader.readInt("day");
+
+                LocalDate date = LocalDate.of(year, month, day);
+                if (type == LOCAL_DATE_TYPE) {
+                    ReflectionUtils.setField(objectField, obj, date);
+                }
+                else {
+                    LocalTime time = LocalTime.of(
+                        reader.readInt("hour"),
+                        reader.readInt("minute"),
+                        reader.readInt("second"),
+                        reader.readInt("nano")
+                    );
+                    LocalDateTime dateTime = LocalDateTime.of(date, time);
+                    ReflectionUtils.setField(objectField, obj, dateTime);
+                }
+            }
+        }
+    }
+
+    /**
+     * Discovery SPI, that makes a node stop sending messages when {@code block} is set to {@code true}.
+     */
+    private static class BlockedDiscoverySpi extends TcpDiscoverySpi {
```

Comment:
can we reuse it from org.apache.ignite.internal.processors.cache.binary.ClientReconnectTest.BlockedDiscoverySpi ?